### ISSUE TITLE
🐛 ClearCookie does not remove cookie

### ISSUE THREAD

**Issue by ruralcoder** [gofiber/fiber#1127]
**Fiber version**
v2

**Issue description**
 ClearCookie does not remove cookie

**Code snippet**

I have 2 routes.

1. a POST Sign in route to set the cookie
```go
cookie := new(fiber.Cookie)
cookie.Name = "JWT"
cookie.Value = "....."
cookie.Domain = "mydomain.com"
cookie.HTTPOnly = true
cookie.Expires = expires

c.Cookie(cookie)

```
and second route is a POST for sign out which removes the cookie

```
c.ClearCookie("JWT")
```

In chrome, the cookie continues to exist after Sign Out and user able to make calls to restricted routes.
Chrome Version `87.0.4280.141 (Official Build) (x86_64)`

**Comment by kiyonlin** [comment]
Cookie is stored in browsers unless it's expired. You can
1. delete it in frontend
2. set the same cookie with an expired time

**Comment by ruralcoder** [comment]
Apart from the fact that `c.ClearCookie("JWT")` should be functional, the client side javascript cannot remove it because of the `HTTPOnly = true`.

I also tried to change the expiration time, but no luck.

**Comment by kiyonlin** [comment]
I wrote a demo and it worked. Please try it and tell me if it works on your side.
```go
package main

import (
	"time"

	"github.com/gofiber/fiber/v2"
)

func main() {
	app := fiber.New()

	app.Get("/", func(c *fiber.Ctx) error {
		cookie := new(fiber.Cookie)
		cookie.Name = "JWT"
		cookie.Value = "token"
		cookie.HTTPOnly = true
		cookie.Expires = time.Now().Add(time.Hour)
		c.Cookie(cookie)
		return c.SendString("set cookie")
	})

	app.Get("/clear", func(c *fiber.Ctx) error {
		c.ClearCookie("JWT")
		return c.SendString("clear cookie")
	})

	_ = app.Listen(":3000")
}
```

**Comment by ruralcoder** [comment]
I can get it to work with

```
cookie := new(fiber.Cookie)
cookie.Name = settingCookieName
cookie.Value = "deleted"
cookie.Domain = "mydomain.com"
cookie.HTTPOnly = true
cookie.Expires = time.Now().Add(3 * time.Second)
```

and the browser accepts it, and wipes it after 3 seconds.

It unblocks me, however,  `c.ClearCookie` still does not work for me.

**Comment by kiyonlin** [comment]
Could you try `cookie.Expires = time.Now().Add(-3 * time.Second)`?

And `c.ClearCookie` internally sets cookie expiration to `CookieExpireDelete = time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)`

**Comment by ruralcoder** [comment]
Cookie was removed with -3

**Comment by ruralcoder** [comment]
Just tried it again with c.ClearCookie("JWT") and still did not work.

**Comment by kiyonlin** [comment]
Tbh, there is no reason that `c.ClearCookie` does not work. I believe maybe somewhere is not right.

**Comment by ruralcoder** [comment]
Here is my code
```
func ClearClaimsCookie(c *fiber.Ctx) {
	// Cookie will eventually expire, but better remove it
	// c.ClearCookie(settingCookieName)

	// cookie := new(fiber.Cookie)
	// cookie.Name = settingCookieName
	// cookie.Value = "deleted"
	// cookie.Domain = settingCookieDomain
	// cookie.HTTPOnly = true
	// cookie.Expires = time.Now().Add(-3 * time.Second)

	// c.Cookie(cookie)

	c.ClearCookie(settingCookieName)
}
```

**Comment by kiyonlin** [comment]
Okay, could you please give us a runnable demo to reproduce this issue?

**Comment by ruralcoder** [comment]
```
package main

import (
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/cors"
)

func CorsHandler() fiber.Handler {

	return cors.New(cors.Config{
		Next:             nil,
		AllowOrigins:     "http://dev.mydomain.com:8080", // Access-Control-Allow-Origin
		AllowMethods:     "GET,POST,HEAD,PUT,DELETE,PATCH",
		AllowHeaders:     "",
		AllowCredentials: true,
		ExposeHeaders:    "",
		MaxAge:           0,
	})
}

func main() {
	app := fiber.New()
	app.Use(CorsHandler())

	app.Options("*", func(c *fiber.Ctx) error {
		// PreflightResponseCode for crossorigin XHR requests
		return c.SendStatus(204)
	})

	app.Post("/auth/signin", func(c *fiber.Ctx) error {
		expiration := time.Now().Add(time.Hour)
		cookie := new(fiber.Cookie)
		cookie.Name = "JWT"
		cookie.Value = "token"
		cookie.Domain = "mydomain.com"
		cookie.HTTPOnly = true
		cookie.Expires = expiration

		c.Cookie(cookie)

		payload := fiber.Map{
			"jwtexp": expiration.Unix(),
		}

		return c.Status(fiber.StatusOK).JSON(&payload)
	})

	app.Post("/auth/signout", func(c *fiber.Ctx) error {

		// does not work....
		//c.ClearCookie("JWT")

		// works fine...
		cookie := new(fiber.Cookie)
		cookie.Name = "JWT"
		cookie.Value = "deleted"
		cookie.Domain = "mydomain.com"
		cookie.HTTPOnly = true
		cookie.Expires = time.Now().Add(-3 * time.Second)
		c.Cookie(cookie)

		return c.SendString("clear cookie")
	})

	app.Get("/account/me", func(c *fiber.Ctx) error {

		jwtString := c.Cookies("JWT")

		if jwtString == "" {
			return c.Status(fiber.StatusUnauthorized).SendString("Opps!")
		}

		profile := fiber.Map{
			"name":  "Stephen",
			"email": "stephen@mydomain.com",
		}

		return c.Status(fiber.StatusOK).JSON(&profile)
	})

	_ = app.Listen(":3000")
}
```

**Comment by ruralcoder** [comment]
I think I know what it is

When I do it, curl returns me
```Set-Cookie: JWT=deleted; expires=Fri, 22 Jan 2021 09:53:25 GMT; domain=mydomain.com; path=/; HttpOnly; SameSite=Lax```

which includes the domain `domain=mydomain.com`

So how would ClearCookie know the domain if it's not a parameter and perhaps it does not use what it knows from the cookie.

curl shows me that when using ClearCookie I get
```
Set-Cookie: JWT=; expires=Tue, 10 Nov 2009 23:00:00 GMT
```

**Comment by ruralcoder** [comment]
Perhaps there is more information behind the scenes

`c.fasthttp.Request.Header.Cookie("JWT")`

But I don't have access as fasthttp is not exported.

**Comment by ruralcoder** [comment]
I think I found another issue...

After running the cookie with and without domain, the browser somehow remembers both tokens with the same key.  So now when Chrome sends the http get /account/me it sends 2 JTW cookies

from Chrome.
```
Cookie: JWT=token; JWT=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlcyI3IiIsImV4cCI6MTYxMTMxMTI1MiwiaXNzIjoiaGtteSIsInN1YiI6IjYwMDdhNzI4YzYzYjdjYmQwZjgyZmJiMCJ9.roQMiZdQcT4PsaJoN-nxZCjZ7BYYOJA1QsnF_sykXm4
```
Never seen this before, same key. twice. Also weird it does not show up in the Application / Cookies for me to delete. I will try to reset my cookie cache, but wanted to share before I do so.

The issue is 
```
jwtString := c.Cookies(settingCookieName)
```

Only returns the first one which is `token`

Not sure what to expect in this case, but this breaks my app until the user resets the cookies.

**Comment by ruralcoder** [comment]
Wiping the browser cookie cache fixed the issue. But the concern of the code hitting this in production....

Perhaps `c.Cookies("JWT")` returns an array.

**Comment by kiyonlin** [comment]
> Perhaps there is more information behind the scenes
> 
> 
> 
> `c.fasthttp.Request.Header.Cookie("JWT")`
> 
> 
> 
> But I don't have access as fasthttp is not exported.
> 
> 

See `Ctx.Context`

**Comment by kiyonlin** [comment]
> Wiping the browser cookie cache fixed the issue. But the concern of the code hitting this in production....
> 
> Perhaps `c.Cookies("JWT")` returns an array or a map of `key: [{value, domain},...]`

I googled it, and found this [how-to-handle-multiple-cookies-with-the-same-name](https://stackoverflow.com/questions/4056306/how-to-handle-multiple-cookies-with-the-same-name), maybe you can take a look.

**Comment by kiyonlin** [comment]
> So Chrome seems to ignore the fact the cookie came from the same subdomain and thus ignores it.
> 
> Btw, my site is on the primary and the REST api code is running on api.domain.com

You're right, nice catch! I think you can use custom way to delete the cookie. 

I'll submit this issue to fasthttp. Thanks again!

**Comment by Fenny** [comment]
Related to https://github.com/gofiber/fiber/issues/692. The `ClearCookie` fn does not know the specific cookie properties set on the client ( like domain, httponly, secure etc ) and some browsers ( differs between chrome and firefox ) require you to set the same properties when expiring an cookie. Other web frameworks have the same problem ( expresssjs ). We do have an idea to come with an user friendly solution by caching cookie properties per key when cookies are created, so when `ClearCookie` is called it will try to build the correct cookie from cache to perform an successful expire with the correct parameters.

We will explore this idea later this week 👍

**Comment by kiyonlin** [comment]
We'll figure this out in a long term. Close it for now.

**Comment by Destreyf** [comment]
I just ran into this, and fenny is right about this issue also existing on express.

I'm just chiming in with what my solution was as I think it can be useful to others.


```go
package utils

import (
	"time"
	"github.com/gofiber/fiber/v2"
)

func SetCookie(c *fiber.Ctx, name string, value string, expiration time.Time) {
	c.Cookie(buildCookie(name, value, expiration))
}

func ClearCookie(c *fiber.Ctx, name string) {
	c.Cookie(buildCookie(name, "", time.Date(2009, time.November, 10, 23, 0, 0, 0, time.UTC)))
}

func buildCookie(name string, value string, expires time.Time) *fiber.Cookie {
	cookie := new(fiber.Cookie)
	cookie.Name = name
	cookie.Value = value
	cookie.HTTPOnly = true
	cookie.Expires = expires
	cookie.Path = "/api/v1/auth/"
	cookie.Domain = "example.com"

	return cookie
}
```

At login I then call my function as 
```go
utils.SetCookie(c, "refresh-token", token, expiration)
```
and to clear it's just 
```go
utils.ClearCookie(c, "refresh-token")
```

This obviously locks the calls to a single endpoint/domain but for my use case (HTTP Only refresh tokens) it works great, my functions are named more specifically and have expiration computed based on an environment variable.

**Comment by adharshmk96** [comment]
I ran into same issue 

Cookie is set by 
```go
c.Cookie(&fiber.Cookie{
  Name:     "account_token",
  Value:    signedJWT,
  Expires:  time.Now().Add(time.Hour * 24),
  HTTPOnly: true,
  SameSite: "Lax",
})
```

clearing cookie

```go
// doesn't work
c.ClearCookie(viper.GetString("auth.cookie_name"))

// Reponse header
// Set-Cookie: account_token=; expires=Tue, 10 Nov 2009 23:00:00 GMT

// works
c.Cookie(&fiber.Cookie{
  Name:     viper.GetString("auth.cookie_name"),
  Value:    "",
  Expires:  time.Now().Add(-time.Hour),
  HTTPOnly: true,
  SameSite: "Lax",
})

// Response header
// Set-Cookie: account_token=; expires=Fri, 22 Mar 2024 16:05:36 GMT; path=/; HttpOnly; SameSite=Lax
```

**Comment by adharshmk96** [comment]
ended up using this 

```go
func ClearCookies(c *fiber.Ctx, key ...string) {
	for i := range key {
		c.Cookie(&fiber.Cookie{
			Name:    key[i],
			Expires: time.Now().Add(-time.Hour * 24),
			Value:   "",
		})
	}
}
```

and usage

```go
utils.ClearCookies(c, "account_token")
```

**Comment by rofleksey** [comment]
Ticket is closed with issue still not fixed xdd

**Comment by dolfelt** [comment]
I ran into this same issue, and had to do a similar thing. I just made a wrapper around the `c.Cookie` that allowed me to pass what I wanted, then created a `ClearCookie` that used the wrapper with an expiration in the past. 

```go
func SetDomainCookie(c *fiber.Ctx, name, value string, expires time.Time) {
	c.Cookie(&fiber.Cookie{
		Name:     name,
		Path:     "/",
		Domain:   ".example.com",
		Value:    value,
		Secure:   true,
		HTTPOnly: true,
		Expires:  expires,
		SameSite: "None",
	})
}

func ClearDomainCookie(c *fiber.Ctx, name ...string) {
	for i := range name {
		SetDomainCookie(c, name[i], "", time.Now().Add(-48*time.Hour))
	}
}
```

From digging into the code, it looks like this is a limitation of fasthttp.

```go
// DelClientCookie instructs the client to remove the given cookie.
// This doesn't work for a cookie with specific domain or path,
// you should delete it manually like:
//
//	c := AcquireCookie()
//	c.SetKey(key)
//	c.SetDomain("example.com")
//	c.SetPath("/path")
//	c.SetExpire(CookieExpireDelete)
//	h.SetCookie(c)
//	ReleaseCookie(c)
//
// Use DelCookie if you want just removing the cookie from response header.
func (h *ResponseHeader) DelClientCookie(key string) {
	h.DelCookie(key)

	c := AcquireCookie()
	c.SetKey(key)
	c.SetExpire(CookieExpireDelete)
	h.SetCookie(c)
	ReleaseCookie(c)
}
```
